Make sure values are defined

hpinkos opened the pull request:
These are two possible questionable spots that coverity found

- `GroundPrimitives` `updateAndQueueCommands`: `boundingVolumes` has to be defined so we can index into it for the commands
- `Model` `createTexture`: make sure texture is defined before trying to use it

cesium-concierge commented:
@hpinkos, thanks for the pull request! Maintainers, we have a signed CLA from @hpinkos, so you can review this at any time.

I noticed that [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CHANGES.md) has not been updated. If this change updates the public API in any way, fixes a bug, or makes any non-trivial update, please add a bullet point to `CHANGES.md` and comment on this pull request so we know it was updated. For more info, see the [Pull Request Guidelines]( https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#pull-request-guidelines).



__I am a bot who helps you make Cesium awesome!__ Thanks again.

mramato commented on Source/Scene/GroundPrimitive.js:
```diff
@@ -599,6 +599,8 @@ define([
         } else if (frameState.mode !== SceneMode.SCENE3D && defined(groundPrimitive._boundingVolumes2D)) {
             boundingVolumes = groundPrimitive._boundingVolumes2D;
+        } else {
+            return;
```
Can this else block ever actually get hit in normal execution?  Would it be better to throw a RuntimeError here instead of silently failing?

hpinkos replied:
@bagnell ?

mramato commented on Source/Scene/Model.js:
```diff
@@ -2306,16 +2306,17 @@ define([
-
-        model._rendererResources.textures[gltfTexture.id] = tx;
-        model._texturesByteLength += tx.sizeInBytes;
+        if (defined(tx)) {
```
This looks like another case that should never happen or a place we should have a Runtime or Developer error.

hpinkos replied:
@lilleyse ?

lilleyse replied:
Yeah this should be a Runtime error.

hpinkos commented:
@mramato ready

bagnell commented on Source/Scene/GroundPrimitive.js:
```diff
@@ -599,6 +601,8 @@ define([
         } else if (frameState.mode !== SceneMode.SCENE3D && defined(groundPrimitive._boundingVolumes2D)) {
             boundingVolumes = groundPrimitive._boundingVolumes2D;
+        } else {
+            throw new RuntimeError('Failed to compute ground primitive bounding volumes');
```
This should never be hit. The only reason `_boundingVolumes2D` won't be defined is when the scene was constructed with `scene3DOnly : true`. If this is hit, there is a bug elsewhere since the scene mode shouldn't be able to change.

hpinkos replied:
So should this be a developer error?

bagnell replied:
I would remove the error and the second condition to make it just an `else`.

mramato commented on Source/Scene/PrimitivePipeline.js:
```diff
@@ -272,17 +272,16 @@ define([
-        var pickOffsets;
-        if (parameters.createPickOffsets && geometries.length > 0) {
-            pickOffsets = createInstancePickOffsets(instances, geometries);
+            if (parameters.createPickOffsets && geometries.length > 0) {
```
You can get rid of `&& geometries.length > 0` and move this into the above if check.

hpinkos replied:
Ah yeah, thanks

hpinkos commented:
@mramato ready

hpinkos commented:
Don't know why travis failed but it doesn't look related to the changes I made

hpinkos commented:
This is ready.  Can someone review and merge?

pjcozzi commented on Source/Scene/GroundPrimitive.js:
```diff
@@ -16,6 +16,7 @@ define([
         '../Core/Math',
         '../Core/OrientedBoundingBox',
         '../Core/Rectangle',
+        '../Core/RuntimeError',
```
Is this used?

pjcozzi commented on Source/Scene/Model.js:
```diff
@@ -2306,14 +2306,18 @@ define([
 
-            if (mipmap) {
-                tx.generateMipmap();
+                if (mipmap) {
```
Are you sure about this?  I'm pretty sure it removes the ability to mipmap textures that are defined by buffers, not images.

hpinkos replied:
This is the same logic that's in master.  You're saying this be moved out of the if block instead?

hpinkos commented:
@pjcozzi  I moved the mipmap logic out of that if block since it did look like it was supposed to work with both buffer and image textures.  I added a comment to CHANGES about this since it was only running for image textures before.

Anything else?

pjcozzi commented on Source/Scene/Model.js:
```diff
@@ -2307,11 +2307,15 @@ define([
-            if (mipmap) {
-                tx.generateMipmap();
-            }
+        if (!defined(tx)) {
```
Is this needed?  Could we just have

```
if (defined(tx) && mipmap) {
    tx.generateMipmap();
}
```

below?

hpinkos replied:
This is needed because we're using the `tx` value below.  So it seems like we're expecting it to be defined.
```
        model._rendererResources.textures[gltfTexture.id] = tx;
        model._texturesByteLength += tx.sizeInBytes;
```

pjcozzi replied:
Maybe I am missing something - couldn't we just put a `defined(tx)` check below or even duplicate the code in both block above?  I am just trying to avoid the random `RuntimeError` to keep Coverity happy.

hpinkos replied:
@pjcozzi apparently I had it that way originally but @lilleyse said it should be a runtime error.  I changed it back to check defined

pjcozzi commented:
> @pjcozzi I moved the mipmap logic out of that if block since it did look like it was supposed to work with both buffer and image textures. I added a comment to CHANGES about this since it was only running for image textures before.

Ah yes, I misread the diff.

@lilleyse can you confirm that we previously had a bug here where we would not generate mipmaps for textures stored in buffers?  Seems like an obvious mistake that would have come up in practice?  Likewise, it looks like we don't resize to be a power of two, which would now be required.

See https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Scene/Model.js#L2276-L2315

lilleyse commented:
@pjcozzi Yeah, it does looks like `mipmaps` were not generated for buffer textures, however this would have only affected glTF 2.0 models. `npot` will need to be checked there too and the texture resized.

pjcozzi commented:
> @pjcozzi Yeah, it does looks like mipmaps were not generated for buffer textures, however this would have only affected glTF 2.0 models. npot will need to be checked there too and the texture resized.

Yes, looks like this was introduced with glTF 2.0, not the binary glTF extension.

Submitted #5905

hpinkos commented:
@pjcozzi do you want me to revert those changes to `Model` so someone fix the mipmapping in a separate PR?

pjcozzi commented:
> @pjcozzi do you want me to revert those changes to Model so someone fix the mipmapping in a separate PR?

Yes, just please move the mipmap call back to the `else` block.

hpinkos commented:
@pjcozzi this is ready now

pjcozzi commented:
Thanks!
